src/TaskRunner.js

```diff
@@ -44,34 +58,53 @@ export default class TaskRunner {
       }
     }
 
-    return Promise.all(
-      tasks.map((task) => {
-        const enqueue = async () => {
-          let result;
+    const limit = pLimit(concurrency);
+    const scheduledTasks = [];
 
-          try {
-            result = await this.runTask(task);
-          } catch (error) {
-            result = { error };
+    for (const file of files) {
+      const enqueue = async (task) => {
+        let taskResult;
+
+        try {
+          taskResult = await this.runTask(task);
+        } catch (error) {
+          taskResult = { error };
+        }
+
+        if (cache.isEnabled() && !taskResult.error) {
+          taskResult = await cache.store(task, taskResult).then(
+            () => taskResult,
+            () => taskResult
+          );
+        }
+
+        task.callback(taskResult);
+
+        return taskResult;
+      };
+
+      scheduledTasks.push(
+        limit(() => {
+          const task = taskGenerator(file).next().value;
+
+          if (!task) {
+            // Something went wrong, for example the `cacheKeys` option throw an error
+            return Promise.resolve();
           }
 
-          if (this.cache.isEnabled() && !result.error) {
-            return this.cache.store(task, result).then(
-              () => result,
-              () => result
+          if (cache.isEnabled()) {
+            return cache.get(task).then(
+              (taskResult) => task.callback(taskResult),
+              () => enqueue(task)
             );
           }
 
-          return result;
-        };
-
-        if (this.cache.isEnabled()) {
-          return this.cache.get(task).then((data) => data, enqueue);
-        }
+          return enqueue(task);
```

Comment:
I don't think `plimit` makes any guarantee about the order of execution of the enqueued functions. I think this means that `enqueue(task)` can be called in any order and therefore `task.callback(taskResult)` can be, too.

This seems like it has the potential to introduce non-determinism whenever parallelism is enabled. That is, if I run my webpack build twice, I expect it to produce the same outputs. This change seems to break that guarantee whenever `extractComments` is on since comment extraction depends on processing the results in a particular order.

It is important that this plugin is deterministic because nondeterministic builds needlessly invalidate long-term caches.

Reply:
Yes, we need sort properties of `allExtractedComments` here https://github.com/webpack-contrib/terser-webpack-plugin/blob/master/src/index.js#L509

Reply:
I will fix it in near future

Reply:
Hm, not sure problem exists:
- https://github.com/webpack-contrib/terser-webpack-plugin/blob/master/src/index.js#L511

We sort comments inside each file, so output is deterministic

- https://github.com/webpack/webpack/blob/bd8c74c0bba3a7732e51a763bc291bd76267e7f7/lib/stats/DefaultStatsPresetPlugin.js#L166

Webpack sort assets by default

Reply:
Ok, I think you're right. Since comments are sorted and then de-duped after being collected in `allExtractedComments`, I think the output is deterministic.